# IEO Quizzes

## Quiz 1

* Started on: Sunday, 7 April 2019, 8:36 PM
* Completed on: Sunday, 7 April 2019, 8:37 PM
* Time taken: 57 secs

### Question 1. High-throughput sequencing of nucleic acids

* a. allows one to partially profile the DNA, RNA and protein content of the cell.
* **b.** *allows one to uniformly profile the DNA and RNA content of the cell.*
* c. allows one to uniformly profile the DNA, RNA, protein and metabolite content of the cell.
* d. allows one to uniformly profile the DNA, RNA and protein content of the cell.
* e. allows one to partially profile the DNA and RNA content of the cell.

### Question 2. In a SummarizedExperiment object

* a. Phenotypic variables in the column data match sample names in the assay data.
* b. Row names in the assay data match the row names in the column data.
* **c.** *Sample column names in the assay data match the sample row names in the column data.*
* d. Phenotypic variables in the sample annotation match feature names in the assay data.
* e. Sample column names in the assay data match the sample row names in the row data.

### Question 3. The detection of RNA expression in whole blood through a high-throughput sequencing experiment depends on

* a. Only on how deep we sequence.
* b. The sample preparation protocol and how large is the sequencer.
* **c.** *The sample preparation protocol and how deep we sequence.*
* d. On the number of samples we sequence and how deep we sequence.
* e. Only on the sample preparation protocol.

### Question 4. A SummarizedExperiment object

* a. follows a matrix-like organisation with samples associated with rows and features with columns.
* **b.** follows a matrix-like organisation with features associated with rows and samples with columns.
* c. organises the data in a manner that is completely different to any other data structure.
* d. follows a vector-like organisation with features associated with rows and samples with columns.
* e. follows a vector-like organisation with samples associated with rows and features with columns.

### Question 5. When doing exome sequencing, the obtained reads should only map to

* a. known (protein) coding regions and promoter regions.
* b. known (protein) coding regions and introns.
* c. known and unknown (protein) coding regions.
* d. known introns.
* **e.** *known (protein) coding regions.* ï¿¼

## Quiz 2

* Started on: Monday, 15 April 2019, 9:00 AM
* Completed on: Monday, 15 April 2019, 9:03 AM
* Time taken: 3 mins 15 secs

### Question 1. The so-called batch effect precludes distinguishing the source of variation that drives the observed results

* **a.** *when is confounded with the outcome of interest in our experiment.*
* b. when the outcome of interest in our experiment is unclear.
* c. when is not confounded with the outcome of interest in our experiment.
* d. always.
* e. only when the data is published in high-impact journals.

### Question 2. RNA-seq data

* a. require a normalization pre-processing step due to the expression properties of the genes.
* b. require a normalization pre-processing step when most genes are not differentially expressed.
* c. require a normalization pre-processing step when most genes are differentially expressed.
* **d.** *require a normalization pre-processing step due to the expression properties of the samples.*
* e. do not require a normalization pre-processing step because are very accurate.

### Question 3. Estimating surrogate variables, with a method such as SVA, can help to

* a. adjust for the batch indicator variable.
* b. adjust for hidden (latent) non-biological factors only.
* **c.** *adjust for hidden (latent) biological and non-biological factors.*
* d. estimate a batch indicator variable.
* e. adjust for hidden (latent) biological factors only.

### Question 4. We read in an article that 30 genes were called differentially expressed (DE) at a FDR level of 10%. This means that

* a. only 30 genes had p-values smaller than 0.1.
* b. only 30 genes had p-values such that one could expect a maximum of 10% of true discoveries among them.
* c. only 30 genes had p-values larger than 0.1.
* d. only 30 genes had p-values smaller than 0.01.
* **e.** *only 30 genes had p-values such that one could expect a maximum of 10% of false discoveries among them.*

### Question 5. In a volcano plot...

* a. most significant gene expression changes are at both ends of the y-axis.
* **b.** *most significant gene expression changes are at the top of the y-axis.*
* c. most significant gene expression changes are at both ends of the x-axis.
* d. most significant gene expression changes are at the bottom of the y-axis.
* e. most significant gene expression changes are always highlighted in red.

## Quizz 3

* Started on: Monday, 13 May 2019, 7:24 AM
* Completed on: Monday, 13 May 2019, 7:26 AM
* Time taken: 1 min 56 secs

### Question 1. The design matrix of a linear model has:

* a. as many rows as variables of the model and as many columns as samples.
* b. as many rows as samples and as many columns as variables of model.
* **c.** *as many rows as samples and as many columns as coefficients of the model.*
* d. as many rows as coefficients of the model and as many columns as samples.
* e. as many rows as genes and as many columns as samples.

### Question 2. Under the null hypothesis of no-differential expression, the distribution of p-values calculated for all the genes should be:

* a. Independent.
* b. Skewed to the left.
* c. Normal (Gaussian).
* **d.** *Uniform.*
* e. Skewed to the right.

### Question 3. We have to analyze a RNA-seq data set of blood samples from patients before and after being treated with a specific drug. Therefore, we have one pair of samples per individual and we want to use a paired design. In a first exploratory analysis, we observe in an MDS plot an effect of the sex among the patients. Should be include sex as explanatory factor in the design matrix employed to conduct the differential expression analysis?

* a. Yes, because sex could affect the response of the patient to the drug.
* **b.** *No, because it's implicitly adjusted when comparing samples in pairs from the same individual.*
* c. No, because it gets automatically adjusted when using surrogate variable analysis.
* d. Yes, because we should always adjust for sex in any differential expression analysis.
* e. Yes, because according to the MDS plot, a fraction of the variability of the RNA-seq data can be attributed to the sex of the patients.

### Question 4. What is it measuring the first coefficient after the intercept in a design matrix that follows the reference model for a differential expression analysis with one binary explanatory variable?

* **a.** *The log fold-change between values of the explanatory variable associated with the coefficient.*
* b. The average expression of the explanatory variable.
* c. The fold-change between explanatory variable and the reference variable.
* d. The average expression for one of the values of the explanatory variable.
* e. The fold-change between values of the explanatory variable associated with the coefficient.

### Question 5. Functional annotations on genes are not...

* a. dynamic, annotations do not change over time.
* **b.** *evenly covering all organisms, some organisms have more annotations than others.*
* c. partial, there are annotations for every gene.
* d. noisy, annotations are always accurate.
* e. essential to downstream bioinformatic analyses.